[23.0] Test validators for optional parameters

ElectronicBlueberry opened the pull request:
Potential fix for optional text parameters not marked as optional.

From the UI/UX working group chat:

> Unlike other types of parameters, `type="text"` parameters are always optional, and tool author need to restrict the input with validator elements

![image](https://user-images.githubusercontent.com/44241786/216291427-1c04bb89-bcc3-410f-9ef5-29ef006cb212.png)

https://github.com/galaxyproject/tools-iuc/blob/main/tools/bcftools/bcftools_isec.xml#L55-L58

-> nfiles allows empty values, but is displayed as required.

Draft PR to check how the tool tests react to this

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ElectronicBlueberry committed:
test validators for optional parameters

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -327,6 +327,15 @@ def __init__(self, tool, input_source):
         super().__init__(tool, input_source)
         self.optional = input_source.get_bool("optional", False)
+
+        if not self.optional:
```
```suggestion
        if not self.optional and self.type == "text":
```
since other parameter types in the class hierarchy inherit from `TextToolParameter`

ElectronicBlueberry committed:
check type

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

ElectronicBlueberry committed:
add validator to optional test

ElectronicBlueberry committed:
fix inline test

mvdbeek committed:
Cast `None` value to `""`

for inferred optional text parameters, for old tools.
This maintains templating behavior for exisiting tools.

mvdbeek committed:
Fix type annotation for ToolParameterValueWrapper
